editor/src/messages/portfolio/document/document_message_handler.rs

```diff
@@ -551,13 +552,40 @@ impl MessageHandler<DocumentMessage, (u64, &InputPreprocessorMessageHandler, &Pe
 					.into(),
 				);
 			}
-			NudgeSelectedLayers { delta_x, delta_y } => {
+			NudgeResizeSelectedLayer { delta_x, delta_y, alt, ctrl } => {
 				self.backup(responses);
+				let keyboard = &ipp.keyboard;
+				let is_scale = keyboard.key(alt);
+				let is_bottom_right = !keyboard.key(ctrl);
 				for path in self.selected_layers().map(|path| path.to_vec()) {
-					let operation = DocumentOperation::TransformLayerInViewport {
-						path,
-						transform: DAffine2::from_translation((delta_x, delta_y).into()).to_cols_array(),
-					};
+					let mut transform_matrix = DAffine2::from_translation((delta_x, delta_y).into()).to_cols_array();
+					if is_scale {
+						if let Ok(Some(existing_bounds)) = self.document_legacy.viewport_bounding_box(&path, &render_data) {
+							let scale_size_x = if is_bottom_right {
+								(existing_bounds[0].x - existing_bounds[1].x - delta_x) / (existing_bounds[0].x - existing_bounds[1].x)
+							} else {
+								(existing_bounds[0].x - existing_bounds[1].x + delta_x) / (existing_bounds[0].x - existing_bounds[1].x)
+							};
+							let scale_size_y = if is_bottom_right {
+								(existing_bounds[0].y - existing_bounds[1].y - delta_y) / (existing_bounds[0].y - existing_bounds[1].y)
+							} else {
+								(existing_bounds[0].y - existing_bounds[1].y + delta_y) / (existing_bounds[0].y - existing_bounds[1].y)
+							};
+							let offset = if is_bottom_right {
+								DAffine2::from_translation((-existing_bounds[0].x, -existing_bounds[0].y).into())
+							} else {
+								DAffine2::from_translation((-existing_bounds[1].x, -existing_bounds[1].y).into())
+							};
+							let scale = DAffine2::from_scale((scale_size_x, scale_size_y).into());
+							let offset_back = if is_bottom_right {
+								DAffine2::from_translation((existing_bounds[0].x, existing_bounds[0].y).into())
+							} else {
+								DAffine2::from_translation((existing_bounds[1].x, existing_bounds[1].y).into())
+							};
+							transform_matrix = (offset_back * scale * offset).to_cols_array();
+						}
```

Comment:
I'm finding this is extremely difficult to read. Could you please start by adding some breathing room with blank lines in all the places they can be used to provide separation between different concepts?

And then here are some general tips you can aim to employ which can reduce the complexity of this code:

- Avoid repeating the `if is_bottom_right` test four separate times.
- By returning a tuple from inside an if statement, then destructuring it outside, you can handle the X and Y values together.
- There is a high degree of code duplication. In your `let scale_size_x = if ... else ...;` and `let scale_size_y = if ... else ...;` expressions, the **only** difference is the `-` and `+` character between the if and else conditions. You should only need to write out that long math expression once and feed a variable that determined the `-` or `+` value.
- `DAffine2::from_scale((scale_size_x, scale_size_y).into());` is equivalent to `DAffine2::from_translation(-existing_bounds[0])`.
- Destructure `existing_bounds` into `[existing_top_left, existing_bottom_right]` for clarity so you can avoid writing `[0]` and `[1]`.
- Reverse the order of `let offset = ...` and `let scale = ...` so you are handling scale separately from offset instead of interweaving those concepts.
- Those are all the tips I can name for now, but try your very best to use equivalent logic to make this code as simple and readable as possible, please.